### PR TITLE
S3

### PR DESCRIPTION
**Pull request by zaibon**

### CODE REVIEW

File: templates/vm/vm.py [hunk 1]
```diff
@@ -79,23 +87,38 @@ def install(self):
...
+            {'name': 'test',
+             'type': 'default'
             }]
         }
 
         image, _, version = self.data['image'].partition(':')
         if image == 'zero-os':
-            version = version or 'master'
-            vm_data['ipxeUrl'] = IPXEURL.format(version)
+            version = version or 'development'
```
Comment (zaibon) [inline templates/vm/vm.py]: should this be `development` ?

File: templates/s3/schema.capnp [hunk 2]
```diff
@@ -0,0 +1,32 @@
+@0xf35b28a27d2d7444;
+
+struct Schema {
```
Comment (zaibon) [inline templates/s3/schema.capnp]: are all the fields mandatory ?
Please add some doc next to the fields and try to put some usable default value where possible

File: templates/s3/s3.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+    def _get_zrobot(self, name, url):
+        j.clients.zrobot.get(name, data={'url': url})
+        return j.clients.zrobot.robots[name]
+
+    def validate(self):
+        if self.data['parityShards'] > self.data['dataShards']:
+            raise ValueError('parityShards must be equal to or less than dataShards')
+
+        capacity = j.clients.grid_capacity.get(interactive=False)
+        self._nodes = json.loads(capacity.api.ListCapacity(query_params={'farmer': self.data['farmerIyoOrg']})[1].content.decode('utf-8'))
```
Comment (zaibon) [inline templates/s3/s3.py]:
```python
resp = capacity.api.ListCapacity(query_params={'farmer': self.data['farmerIyoOrg']})
resp.raise_for_status()
self._nodes  = resp.json()
```

File: templates/s3/s3.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+                if available:
+                    best_node[storage_key] = best_node[storage_key] - self.data['storageSize']
+                    self.data['namespaces'].append({'name': namespace.name, 'url': best_node['robot_address'], 'node': best_node['node_id']})
+                    return namespace, next_index
+                namespace.delete()
+            if next_index == index:
+                raise RuntimeError('Looped all nodes and could not find a suitable node')
+            final_index = next_index
+        
+    def install(self):
```
Comment (zaibon) [inline templates/s3/s3.py]: would be nice to have a doc string and/or comments that explain the flow, this function is pretty big so hard to understand

File: templates/s3/s3.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        minio.schedule_action('start').wait(die=True)
+        port = minio.schedule_action('node_port').wait(die=True).result
+        self.data['minioUrl'] = 'http://{}:{}'.format(member['config']['ipAssignments'][0], port)
+
+    def uninstall(self):
+        for namespace in self.data['namespaces']:
+            robot = self._get_zrobot(namespace['node'], namespace['url'])
+            ns = robot.services.get(template_uid=NS_TEMPLATE_UID, name=namespace['name'])
+            ns.schedule_action('uninstall').wait(die=True)
+            self.data['namespaces'].remove(namespace)
```
Comment (zaibon) [inline templates/s3/s3.py]: you need to also delete the services on the nodes after they are uninstalled

File: templates/vm/vm.py (hunk 1 above)
Comment (serboctor) [reply templates/vm/vm.py]: @zaibon master doesn't have the new ztid feature in the kernel args

Comment (zaibon) [reply templates/vm/vm.py]: yeah that means we have to draft a new release for 0-os to roll out this new feature, but having development hardcoded I don't think that's a good idea

File: templates/s3/s3.py [hunk 6]
```diff
@@ -52,30 +55,42 @@ def _create_namespace(self, index, storage_key, password):
...
                     template_uid=NS_TEMPLATE_UID, data=data)
-                available = namespace.schedule_action('install', args={'if_available': True}).wait(die=True).result
-                if available:
+
+                task = namespace.schedule_action('install').wait()
+                if task.eco:
+                    if task.eco.exceptionclassname == 'NoNamespaceAvailability':
+                        namespace.delete()
+                    else:
+                        raise RuntimeError(task.eco.errormessage)
```
Comment (zaibon) [inline templates/s3/s3.py]: to be checked, but i think an eco can be raised directly, it is a child of Exception. At least it should be

**Pull request merged**